src/hotspot/share/classfile/moduleEntry.cpp

```diff
@@ -467,27 +484,58 @@ void ModuleEntry::init_as_archived_entry() {
   ArchivePtrMarker::mark_pointer((address*)&_location);
 }
 
-void ModuleEntry::init_archived_oops() {
+void ModuleEntry::update_oops_in_archived_module(int root_oop_index) {
   assert(DumpSharedSpaces, "static dump only");
-  oop module_obj = module();
-  if (module_obj != NULL) {
-    oop m = HeapShared::find_archived_heap_object(module_obj);
-    assert(m != NULL, "sanity");
-    _archived_module_index = HeapShared::append_root(m);
-  }
+  assert(_archived_module_index == -2, "must be set exactly once");
+  assert(root_oop_index >= 0, "sanity");
+
+  _archived_module_index = root_oop_index;
+
   assert(shared_protection_domain() == NULL, "never set during -Xshare:dump");
   // Clear handles and restore at run time. Handles cannot be archived.
   OopHandle null_handle;
   _module = null_handle;
+
+  // For validate_archived_module_entries()
+  DEBUG_ONLY(_num_inited_module_entries++);
+}
+
+char* ModuleEntry::debug_info() const {
```

Comment:
Isn't this just print_on?  or print_string, although print_on(outputStream*) would be more consistent with other metadata.